### PR TITLE
Add `cupy.median`

### PR DESCRIPTION
**Pull request by Harshan01**
For the issue #2305

The implementation still lacks the functionality to calculate median for `float16`, `bool` and `complex` dtypes.

- `float16` and `bool` because of `thrust.sort`.
- `complex` because of `cupy.partition`.

### CODE REVIEW

**Comment by emcastillo** [comment]
Can you compare the time for several array sizes using `cupyx.time.repeat` and paste the outputs here?

**Comment by Harshan01** [comment]
These are the time comparison tests with `warmup = 10` and `n = 100`.

Code:
```Python console
In [1]: import numpy as np                                                                                                                                                                                  

In [2]: import cupy as cp                                                                                                                                                                                   

In [3]: from cupyx.time import repeat                                                                                                                                                                       

In [4]: def compare(): 
    ...:     n = 10 ** np.arange(1, 8) 
    ...:     for i in range(n.shape[0]): 
    ...:         print('\nN = ', n[i]) 
    ...:         for xp in [np, cp]: 
    ...:             a = n[i] * xp.random.rand(n[i]) 
    ...:             times = repeat(xp.median, (a,), n=100) 
    ...:             print(xp.__name__, '\t\t', times.to_str(show_gpu=(xp.__name__ == 'cupy'))) 
    ...:              
    ...:                                                                                                                                                                                                    

In [5]: compare()                                                                                                                                                                                          

N =  10
/mnt/hdd/CuPy/cupy/cupyx/time.py:44: FutureWarning: cupyx.time.repeat is experimental. The interface can change in the future.
  util.experimental('cupyx.time.repeat')
numpy 		 median              :    38.396 us   +/- 1.609 (min:   35.863 / max:   49.754) us
cupy 		 median              :    88.570 us   +/-12.157 (min:   84.357 / max:  195.765) us     91.695 us   +/-12.255 (min:   87.648 / max:  199.680) us

N =  100
numpy 		 median              :    39.131 us   +/- 0.697 (min:   38.352 / max:   45.438) us
cupy 		 median              :    95.871 us   +/-111.136 (min:   80.347 / max: 1197.500) us    101.609 us   +/-138.435 (min:   82.944 / max: 1475.584) us

N =  1000
numpy 		 median              :    40.803 us   +/- 0.951 (min:   39.664 / max:   47.224) us
cupy 		 median              :    95.022 us   +/-44.230 (min:   83.271 / max:  416.415) us    100.869 us   +/-56.756 (min:   73.760 / max:  548.832) us

N =  10000
numpy 		 median              :   125.878 us   +/-24.721 (min:  114.329 / max:  257.740) us
cupy 		 median              :   114.252 us   +/-21.483 (min:  106.817 / max:  229.031) us    319.241 us   +/-49.082 (min:  281.600 / max:  783.296) us

N =  100000
numpy 		 median              :   819.603 us   +/-11.381 (min:  810.497 / max:  869.793) us
cupy 		 median              :   114.833 us   +/-17.819 (min:  107.448 / max:  200.357) us    747.350 us   +/-21.567 (min:  720.896 / max:  806.912) us

N =  1000000
numpy 		 median              :  9177.139 us   +/-148.874 (min: 8982.810 / max: 9887.091) us
cupy 		 median              :   112.109 us   +/- 8.304 (min:  106.868 / max:  164.212) us   3385.560 us   +/-60.704 (min: 3322.880 / max: 3777.536) us

N =  10000000
numpy 		 median              : 112332.566 us   +/-659.415 (min:111274.355 / max:114992.044) us
cupy 		 median              :   123.798 us   +/-21.219 (min:  110.564 / max:  256.301) us  29123.723 us   +/-268.866 (min:28765.184 / max:30496.511) us

```

**Comment by Harshan01** [comment]
Right now, `cupy.median` achieves 3x speedups compared to `numpy.median`. I would like to know typically what speedup is required for CuPy in general as compared to numpy.

File: cupy/core/_routines_statistics.pyx [hunk 1]
```diff
@@ -310,6 +311,57 @@ cdef _nanargmax_func = create_reduction_func(
...
+        part = a.copy()
+
+    if axis is None:
+        part = part.ravel()
+        part.partition(kth)
+    else:
+        part.partition(kth, axis=axis)
+
+    if part.shape == ():
+        return part.item()
```
Comment (grlee77) [inline cupy/core/_routines_statistics.pyx]: I think this case should just return a 0-dimensional array for consistency with other CuPy functions like `mean`? item() will cause host->device transfer.

Comment (emcastillo) [reply cupy/core/_routines_statistics.pyx]: good catch

**Comment by emcastillo** [comment]
A way to improve performance is to make a custom kernel similar to the partition one, but that it computes the actual median.
That should give one or two orders of magnitude speedup.

**Comment by Harshan01** [comment]
I will start working on it

**Comment by Harshan01** [comment]
@emcastillo I have modified the partition kernel to find median. However, the case where is `data.sort` is not used (the case where the kernel is used) is very slim. So, for large cases, `data.sort` is always used. 

I tried to force the use of kernel for large sizes of arrays. However, since only 1 thread-block is reserved for 1 median computation, it becomes very slow.

Can you suggest in which direction to head towards for solving this problem?

**Comment by emcastillo** [comment]
Do you have performance numbers?

**Comment by Harshan01** [comment]
Yes, the results are very similar to previous results. I suspect because only 1 case in this (N = 1000) uses the partition kernel, rest all cases use `data.sort`.

```Python console
N =  10
/mnt/hdd/CuPy/cupy/cupyx/time.py:44: FutureWarning: cupyx.time.repeat is experimental. The interface can change in the future.
  util.experimental('cupyx.time.repeat')
numpy 		 median              :    37.121 us   +/- 7.642 (min:   34.195 / max:  117.366) us
cupy 		 median              :    89.660 us   +/-42.351 (min:   79.771 / max:  846.141) us     95.779 us   +/-61.241 (min:   57.344 / max:  899.744) us

N =  100
numpy 		 median              :    38.290 us   +/- 9.591 (min:   34.716 / max:  120.646) us
cupy 		 median              :    92.013 us   +/-42.340 (min:   80.789 / max:  899.718) us     97.592 us   +/-63.559 (min:   57.344 / max: 2719.616) us

N =  1000
numpy 		 median              :    47.340 us   +/-13.295 (min:   38.028 / max:  194.218) us
cupy 		 median              :    86.843 us   +/-33.376 (min:   42.569 / max:  736.843) us    312.725 us   +/-55.115 (min:  240.640 / max: 1246.848) us

N =  10000
numpy 		 median              :   155.695 us   +/-38.452 (min:  116.396 / max:  352.022) us
cupy 		 median              :   132.789 us   +/-34.265 (min:  109.944 / max:  370.969) us    280.577 us   +/-32.238 (min:  234.496 / max: 1275.744) us

N =  100000
numpy 		 median              :  1014.378 us   +/-124.227 (min:  915.700 / max: 3486.970) us
cupy 		 median              :   119.723 us   +/-15.517 (min:  106.753 / max:  291.700) us    703.519 us   +/-13.625 (min:  670.720 / max: 1011.712) us

N =  1000000
numpy 		 median              : 10403.370 us   +/-226.376 (min:10185.394 / max:15699.032) us
cupy 		 median              :   112.896 us   +/-10.135 (min:  107.974 / max:  340.375) us   3467.556 us   +/-32.791 (min: 3394.432 / max: 4072.352) us
```

**Comment by emcastillo** [comment]
CuPy expects large arrays to benefit from GPU computing so for N<=10.000 results seem acceptable to me. 
@asi1024 do you think performance for this new routine is good?

**Comment by Harshan01** [comment]
I retried the experiment.

```Python console
N =  100000
numpy 		 median              :   737.532 us   +/-28.898 (min:  721.409 / max: 1136.042) us
cupy 		 median              :   109.805 us   +/-12.611 (min:  105.929 / max:  287.610) us    703.741 us   +/-10.898 (min:  679.936 / max: 1012.736) us

N =  1000000
numpy 		 median              :  8713.668 us   +/-142.167 (min: 8519.093 / max:11804.668) us
cupy 		 median              :   110.158 us   +/-11.709 (min:  106.079 / max:  269.340) us   3484.940 us   +/-30.610 (min: 3409.824 / max: 3886.080) us
```

Oddly, only N=100000 case has increased from 657.440 us to around 700 us.
But N=1000000 case has decreased from 3544.744 us to around 3480 us.

**Comment by emcastillo** [comment]
I would check the implementation in detail later, but if performance is roughly similar to the previous implementation (The difference seems mostly statistical variability rather than actual improvement), I'd rather go with the original approach as the code is much less and helps maintainability.

File: cupy/core/_routines_statistics.pyx [hunk 2]
```diff
@@ -310,6 +322,234 @@ cdef _nanargmax_func = create_reduction_func(
...
+    # These parameters are determined from the measurement on TITAN X.
+    t = 4
+    sz = 1
+    while sz > 0 and length // sz < max_k + 32 * t:
+        sz //= 2
+    sz *= data.size // length
+
+    # If the array size is small or k is large, we simply sort the array.
+    if length < 32 or sz < 1 or max_k >= 1024:
+        # kth is ignored.
```
Comment (emcastillo) [inline cupy/core/_routines_statistics.pyx:473]: Have you tried to always launch your kernel for smaller array sizes?

Comment (Harshan01) [reply cupy/core/_routines_statistics.pyx:473]: Do you mean to say to launch the kernel for small sizes and then merge them together to fully sort the array?

Comment (emcastillo) [reply cupy/core/_routines_statistics.pyx:473]: well, I guess its ok.

Let's just revert to the original implementation as there is no improvement for this one.
Sorry for making you work this hard.

Comment (Harshan01) [reply cupy/core/_routines_statistics.pyx:473]: No problem. I researched a lot, and it seems that there is no viable parallel implementation for 'selection algorithms' as they are highly linear.

I will revert to the original implementation.

File: cupy/core/_routines_statistics.pyx [hunk 3]
```diff
@@ -5,6 +5,7 @@ from cupy.core._reduction import create_reduction_func
 from cupy.core._reduction import ReductionKernel
 
 from cupy.core cimport _routines_math as _math
+from cupy.core cimport _routines_sorting as _sorting
```
Comment (leofang) [inline cupy/core/_routines_statistics.pyx:8]: Looks like this line can be removed now?

File: cupy/core/_routines_statistics.pyx [hunk 4]
```diff
@@ -310,6 +311,57 @@ cdef _nanargmax_func = create_reduction_func(
     None, _min_max_preamble)
 
 
+cpdef ndarray _median(
+        ndarray a, axis, out, overwrite_input, keepdims):
+
+    if not isinstance(a, ndarray):
+        raise TypeError('Array is not cupy.ndarray')
```
[the remark this answers is not in the document]
Comment (Harshan01) [reply cupy/core/_routines_statistics.pyx:318]: Yes, my bad. I will remove it

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**